ledger/boxtxn_test.go

```diff
@@ -369,3 +371,98 @@ func TestBoxRW(t *testing.T) {
 		dl.txn(withBr)
 	})
 }
+
+// TestBoxAccountData tests that an account's data changes when boxes are created
+func TestBoxAccountData(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	uint64ToArgStr := func(i uint64) string {
+		encoded := make([]byte, 8)
+		binary.BigEndian.PutUint64(encoded, i)
+		return string(encoded)
+	}
+
+	genBalances, addrs, _ := ledgertesting.NewTestGenesis()
+	ledgertesting.TestConsensusRange(t, boxVersion, 0, func(t *testing.T, ver int, cv protocol.ConsensusVersion) {
+		dl := NewDoubleLedger(t, genBalances, cv)
+		defer dl.Close()
+
+		proto := config.Consensus[cv]
+
+		var bufNewLogger bytes.Buffer
+		log := logging.NewLogger()
+		log.SetOutput(&bufNewLogger)
+
+		appIndex := dl.fundedApp(addrs[0], 1_000_000, boxAppSource)
+		call := txntest.Txn{
+			Type:          "appl",
+			Sender:        addrs[0],
+			ApplicationID: appIndex,
+			Boxes:         []transactions.BoxRef{{Index: 0, Name: []byte("x")}, {Index: 0, Name: []byte("y")}},
+		}
+
+		verifyAppSrc := main(`
+txn ApplicationArgs 0
+btoi
+txn Accounts 1
+acct_params_get AcctMinBalance
+assert
+==
+assert
+
+txn ApplicationArgs 1
+btoi
+txn Accounts 1
+acct_params_get AcctTotalBoxes
+assert
+==
+assert
+
+txn ApplicationArgs 2
+btoi
+txn Accounts 1
+acct_params_get AcctTotalBoxBytes
+assert
+==
+assert
+`)
+		verifyAppIndex := dl.fundedApp(addrs[0], 0, verifyAppSrc)
+		verifyAppCall := txntest.Txn{
+			Type:          "appl",
+			Sender:        addrs[0],
+			ApplicationID: verifyAppIndex,
+			Accounts:      []basics.Address{appIndex.Address()},
+		}
+
+		// The app account has no box data initially
+		dl.txn(verifyAppCall.Args(uint64ToArgStr(proto.MinBalance), "\x00", "\x00"))
+
+		dl.txn(call.Args("create", "x", "\x10")) // 16
+
+		// It gets updated when a new box is created
+		dl.txn(verifyAppCall.Args(uint64ToArgStr(proto.MinBalance+proto.BoxFlatMinBalance+17*proto.BoxByteMinBalance), "\x01", "\x11"))
+
+		dl.txn(call.Args("create", "y", "\x05"))
+
+		// And again
+		dl.txn(verifyAppCall.Args(uint64ToArgStr(proto.MinBalance+2*proto.BoxFlatMinBalance+23*proto.BoxByteMinBalance), "\x02", "\x17"))
+
+		// Advance more than 320 rounds, ensure box is still there
+		for i := 0; i < 330; i++ {
+			dl.fullBlock()
+		}
+		time.Sleep(5 * time.Second) // balancesFlushInterval, so commit happens
+		dl.fullBlock(call.Args("check", "x", string(make([]byte, 16))))
+		time.Sleep(100 * time.Millisecond) // give commit time to run, and prune au caches
+		dl.fullBlock(call.Args("check", "x", string(make([]byte, 16))))
+
+		// Still the same after caches are flushed
```

Comment:
I don't follow why this is sure to remove the account object from the lru cache.

Reply:
Apologies, I was using cache as a blanket statement for anything before the DB is hit.

The LRU and the DB actually suffered from the same problem, since they use the same representation for accounts (`persistedAccountData`). So if the account lookup hit either of these, its box fields would be returned as zeros.

Only if the account lookup returns an account from the `accountUpdates` deltas could its box fields have non-zero values. That's because the account is stored as a `ledgercore.AccountData` struct here, which has support for the box fields prior to this PR. And importantly, the box fields populated here would only reflect recent changes to an account's boxes, since the fields were reset to zero as soon as an account is evicted from `accountUpdates`.

Reply:
Yes, that's probably a good idea. I'm not sure if tests at this level use an actual DB or an in-memory one, but in-memory should behave the same for testing purposes.

Flushing the LRU at random intervals during e2e tests might also help guard against any changes in behavior it might introduce.

Reply:
For completeness, https://github.com/algorand/go-algorand/pull/4669 is a related PR attempting to defend against regressions.